### PR TITLE
Replacing LockFreeArray with TBB's concurrent_vector

### PR DESCRIPTION
**Pull request by zhaoguoquan94**
This PR replace current fix size (1024*1024) LockFreeArray with Intel's TBB concurrent_vector. 
#973

### CODE REVIEW

**Comment by zhaoguoquan94** [comment]
Memory usage is now reduced to 347MB. It used to be 619MB (no query executed at all)

virtualbox_kubuntu_26_02_2018_22_45_48

**Comment by poojanilangekar** [comment]
@zhaoguoquan94 A part of your changes from this PR are already merged in the master branch. But I explicitly reverted the lock_free_array changes. So you need to add that back. Also you have mixed multiple other commits in this PR. So here is what I would suggest. Create a new branch, cherrypick changes from this PR and raise a new PR. 

In the future, always use the following commands to rebase a PR:
`git remote add upstream git://github.com/cmu-db/peloton.git`
`git fetch upstream`
`git rebase -i upstream/master`

**Comment by zhaoguoquan94** [comment]
@tcm-marcel 
> What does it make lock free? (before switching to the concurrent vector)

Previously the lock free is just a big std::array of size 1048576. There is an atomic int index and each time the user append something, the atomic index get incremented. 

> Do we still need the LockFreeArray abstraction after switching to the TBB concurrent vector?

Actually we don't need it but I suggest we keep it. As I described here #973 ,there are two use cases but they actually want different functionalities. I remove the LockFreeArray with tbb::concurrent_unordered_map [manager.h](https://github.com/cmu-db/peloton/pull/1168/files#diff-7003eb19b062714b43f6cdf6b1e9ff90R105) because manager don't even insert (append) stuff. It just update the elements at specific index, so I replace it with tbb::concurrent_unordered_map. The reason that I should keep the LockFreeArray abstraction is that tbb::concurrent_vector has more functionalities than we need and documentation is not well formatted.  You need to click many links until you find the correct doc page. We can keep it so that new Peloton developers can use it just by looking at the header.

I agree with @poojanilangekar it's better to create a new PR since part of this PR has been merged into master. I'll close it and create a new PR asap.

File: src/common/container/lock_free_array.cpp [hunk 1]
```diff
@@ -97,41 +93,30 @@ ValueType LOCK_FREE_ARRAY_TYPE::FindValid(const std::size_t &offset,
...
 LOCK_FREE_ARRAY_TEMPLATE_ARGUMENTS
-bool LOCK_FREE_ARRAY_TYPE::IsEmpty() const{
-  return lock_free_array->empty();
-}
+bool LOCK_FREE_ARRAY_TYPE::IsEmpty() const { return lock_free_array.empty(); }
 
 LOCK_FREE_ARRAY_TEMPLATE_ARGUMENTS
-void LOCK_FREE_ARRAY_TYPE::Clear(const ValueType& invalid_value) {
+void LOCK_FREE_ARRAY_TYPE::Clear(const ValueType &invalid_value) {
+  // Set invalid value for all elements
```
[the remark this answers is not in the document]
Comment (zhaoguoquan94) [reply src/common/container/lock_free_array.cpp]: Old LockFreeArray is a fixed size std::array so you cannot clear it. tbb can change the size. You are right, we should just clear it.

**Pull request closed without merging**